### PR TITLE
Log the parsed (and pretty-printed) response body at `LogLevel(-1_000`)

### PR DESCRIPTION
**Pull request by DilumAluthge**
Replaces #130 (closes #130)

TODO:
- [x] Add tests

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -16,16 +17,18 @@ URIs = "5c2747f8-b7ea-4ff2-ba2e-563bfd36b1d4"
 UUIDs = "cf7118a7-6976-5b1a-9a39-7adc72f591a4"
 
 [compat]
-Base64 = "1"
-Dates = "1"
+Base64 = "<0.0.1, 1"
+Dates = "<0.0.1, 1"
```
Comment (devmotion) [inline Project.toml:21]: I think this is not needed - at least it was not needed for 2.2.0 a few days ago:
```suggestion
Base64 = "1"
Dates = "1"
```

File: Project.toml [hunk 2]
```diff
@@ -16,16 +17,18 @@ URIs = "5c2747f8-b7ea-4ff2-ba2e-563bfd36b1d4"
...
 HTTP = "0.9.8, 1"
 JSON3 = "1.5.1"
-Mmap = "1"
+Logging = "<0.0.1, 1"
+Mmap = "<0.0.1, 1"
```
Comment (devmotion) [inline Project.toml:25]:
```suggestion
Logging = "1"
Mmap = "1"
```

File: src/requests.jl [hunk 3]
```diff
@@ -248,25 +248,40 @@ for the FHIR `client`, and parse the JSON response with JSON3 as an object of ty
...
+    # Recall that the default log levels are:
+    #     Error === LogLevel(2_000)
+    #     Warn === LogLevel(1_000)
+    #     Info === LogLevel(0)
+    #     Debug === LogLevel(-1_000)
+    #
+    # Ref: https://docs.julialang.org/en/v1/stdlib/Logging
+
+    # This is really only necessary if you run into errors when parsing the JSON.
+    @logmsg LogLevel(-5_000) "" response_body
```
Comment (devmotion) [inline src/requests.jl]: Maybe mention the function call and add the request to the log message?
```suggestion
    @logmsg LogLevel(-5_000) "FHIRClient.request" response_body path verb
```

File: src/requests.jl [hunk 4]
```diff
@@ -248,25 +248,40 @@ for the FHIR `client`, and parse the JSON response with JSON3 as an object of ty
...
+
+    @logmsg LogLevel(-3_000) "" JSON3.read(response_body)
```
Comment (devmotion) [inline src/requests.jl]: I'm not sure if this should be done, I am a bit worried that this will throw an error. For instance, this is problematic if the response is not in the JSON format. I also wonder what additional benefit this provides over just logging `response_body`?

File: Project.toml [hunk 5]
```diff
@@ -16,16 +17,18 @@ URIs = "5c2747f8-b7ea-4ff2-ba2e-563bfd36b1d4"
...
 SaferIntegers = "2.5.1, 3"
 StructTypes = "1.2.3"
+Test = "<0.0.1, 1"
```
Comment (devmotion) [inline Project.toml:28]: Not needed, I think?
```suggestion
```

File: Project.toml [hunk 6]
```diff
@@ -16,16 +17,18 @@ URIs = "5c2747f8-b7ea-4ff2-ba2e-563bfd36b1d4"
...
 TimeZones = "1.5.2"
 URIs = "1.3"
-UUIDs = "1"
+UUIDs = "<0.0.1, 1"
```
Comment (devmotion) [inline Project.toml:31]:
```suggestion
UUIDs = "1"
```

File: test/runtests.jl [hunk 7]
```diff
@@ -3,9 +3,25 @@ using Test
...
+    # When we run the tests, we want to make sure that none of the `@logmsg`
+    # statements throw any errors.
+    test_logger = Test.TestLogger(;
+        # min_level = Logging.BelowMinLevel,
+        min_level = Logging.LogLevel(-5_000),
+        catch_exceptions = false,
+    );
+    Logging.with_logger(test_logger) do
+        include("unit.jl")
+        include("integration.jl")
```
Comment (devmotion) [inline test/runtests.jl:25]: I think this is missing tests of the logging messages?

File: test/runtests.jl [hunk 8]
```diff
@@ -3,9 +3,25 @@ using Test
 
 import Dates
 import JSON3
+import Logging
 import URIs
 
+# This is necessary to work around https://github.com/JuliaLang/julia/issues/52234
+# Also possibly related: https://github.com/JuliaLang/julia/issues/34037
```
Comment (devmotion) [inline test/runtests.jl:10]: This makes it quite annoying to work with the log-levels proposed in this PR. I think maybe we should only work with log-levels >= -1000: As explained in https://julialogging.github.io/tutorials/logging-basics/#Log-levels log-levels >= -1000 and < 0 are intended for debug logging.

File: Project.toml (hunk 1 above)
Comment (DilumAluthge) [reply Project.toml:21]: So, if there is a package PkgA that depends on FHIRClient, and if PkgA performs any Pkg operations in its test suite, then PkgA's test suite will fail unless we include the `<0.0.1`.

Since we don't know about any private packages out there that depend on FHIRClient, to be safe, we need to include the `<0.0.1`.

File: Project.toml (hunk 2 above)
Comment (DilumAluthge) [reply Project.toml:25]: https://github.com/JuliaHealth/FHIRClient.jl/pull/132/files#r1405237039

File: Project.toml (hunk 6 above)
Comment (DilumAluthge) [reply Project.toml:31]: https://github.com/JuliaHealth/FHIRClient.jl/pull/132/files#r1405237039

File: Project.toml (hunk 5 above)
Comment (DilumAluthge) [reply Project.toml:28]: Currently, there is no real need to have compat entries for stdlibs that are test-dependencies. For example, AutoMerge will not complain about it.

However, the proposal for the future of stdlibs is that all (or almost all) stdlibs will become "upgradeable". This means that in the future, it will become possible for us to release a breaking 2.0 release of the Test.jl stdlib. So, in order to future-proof ourselves against those future breaking releases, we need to start adding compat entries for stdlibs that are test-dependencies.

File: Project.toml (hunk 1 above)
Comment (devmotion) [reply Project.toml:21]: I just noticed that there was another update of https://discourse.julialang.org/t/psa-compat-requirements-in-the-general-registry-are-changing/104958 that mentions this Pkg problem. My comment was based on the fact that we only support Julia >= 1.4 in FHIRClient, and hence based on the previous announcements and the successful registration of release 2.2.0 I assumed that we do not have to add `<0.0.1` in FHIRClient. It seems quite unfortunate and undesirable that apparently it is possible to accidentally break other packages even if FHIRClient itself is completely fine and it's (currently) not possible to notice these problems. I think that most packages that were updated based on the previous announcements are not aware of this fact (most commonly, I've seen people just adding `... = "1"` entries).

Comment (DilumAluthge) [reply Project.toml:21]: Yeah, my thinking now is that we just need to tell everyone to just use `"<0.0.1, 1"`.

Comment (DilumAluthge) [reply Project.toml:21]: We might need to modify AutoMerge so that it fails if someone doesn't have the `<0.0.1` for a stdlib.

File: test/runtests.jl (hunk 7 above)
Comment (DilumAluthge) [reply test/runtests.jl:25]: Yeah, we still need to add those tests.

File: src/requests.jl (hunk 4 above)
Comment (DilumAluthge) [reply src/requests.jl]: The default ConsoleLogger would catch the exception here, but in general, you're right that we shouldn't rely on that.

So let's explicitly make sure that this doesn't throw: #138 

The big benefit of logging this is that `JSON3.read(response_body)` will be pretty-printed to the logs; thus, when reading the logs, it's going to be a lot easier to read `JSON3.read(response_body)` than to read the raw `response_body`.

File: Project.toml [hunk 9]
```diff
@@ -1,13 +1,14 @@
 name = "FHIRClient"
 uuid = "b44d2ca2-8176-4fa9-8684-826e17b2a2da"
 authors = ["Dilum Aluthge", "Rhode Island Quality Institute", "contributors"]
-version = "2.2.0"
+version = "2.3.0"
```
Comment (DilumAluthge) [inline Project.toml]: There are a couple changes on master now, so let's avoid touching the version number here, and then we can bump the version in a separate standalone PR.

```suggestion
version = "2.2.0"
```

File: src/requests.jl (hunk 4 above)
Comment (DilumAluthge) [reply src/requests.jl]: Alright, I've merged #138 into this PR. Now, we will catch any exceptions that JSON3 might throw.

File: test/runtests.jl (hunk 8 above)
Comment (DilumAluthge) [reply test/runtests.jl:10]: Okay, for now, let's stick with LogLevels >= -1000.

https://github.com/JuliaHealth/FHIRClient.jl/pull/155

However, long-term, I think I'll want to revisit this. I think it would be unfortunate if the entire Julia ecosystem were restricted to only working with that narrower range of log levels, and I think it would be better if more Julia packages logged at a wider range of debug log levels, instead of just always defaulting to -1000.

File: test/runtests.jl (hunk 7 above)
Comment (DilumAluthge) [reply test/runtests.jl:25]: Tests added in https://github.com/JuliaHealth/FHIRClient.jl/pull/156 - thank you David!

**Review by devmotion — APPROVED** [review]

**Comment by DilumAluthge** [comment]
@devmotion Can you register a new version?
